calculator/calculator.py

```diff
@@ -1,15 +1,115 @@
+    if type(expression) is not str:
+        raise TypeError("Expression input should be of str type.")
 
+    return _calculate(expression.replace(" ", ""))
 
+def _calculate(expression: str) -> Union[int, float]:
+    if _is_end_of_calculation(expression):
+        return expression
 
+    for _, value in _TOKEN_SPECIFICATION.items():
+        tmp_result = _handle(expression, value)
+        if tmp_result is not None:
+            return _calculate(tmp_result)
+
+    raise ValueError(f'Expression: {expression} is not legal!')
+
+def _one_variable_operation(operation: str, is_negative_legal: bool = False) -> str:
+    if is_negative_legal is False:
+        return f'(?P<NUMBER>{_NUMBER_REGEX}){operation}'
+    else:
+        return f'(?P<NUMBER>{_NUMBER_NEGATIVE_OPT_REGEX}){operation}'
+
+def _two_variable_operation(operation: str, is_negative_legal: bool = False) -> str:
+    if is_negative_legal is False:
+        return f'(?P<FIRST_NUMBER>{_NUMBER_REGEX}){operation}' \
+               f'(?P<SECOND_NUMBER>{_NUMBER_REGEX})'
+    else:
+        return f'(?P<FIRST_NUMBER>{_NUMBER_NEGATIVE_OPT_REGEX}){operation}' \
+               f'(?P<SECOND_NUMBER>{_NUMBER_NEGATIVE_OPT_REGEX})'
+
+def _is_end_of_calculation(expression: str) -> bool:
+    return re.fullmatch(f'{_NUMBER_NEGATIVE_OPT_REGEX}', expression) is not None
+
+def _handle(expression: str, operation_details) -> str:
+    token = re.search(operation_details[OPERATION_REGEX_INDEX], expression)
+    if token:
+        return re.sub(token.re, operation_details[OPERATION_HANDLER_INDEX](token), expression, 1)
+
+def _handle_parenthesis(token) -> str:
+    return str(_calculate(token.group("EXTRACTED_EXPRESSION")))
+
+def _handle_factorial(token) -> str:
+    return str(math.factorial(float(token.group("NUMBER"))))
+
+def _handle_negate(token) -> str:
+    return str(-1 * float(token.group("NUMBER")))
+
+def _handle_max_min_average(token) -> str:
+    if token.group("MAX"):
+        return str(max(float(token.group("FIRST_NUMBER")), float(token.group("SECOND_NUMBER"))))
+    elif token.group("MIN"):
+        return str(min(float(token.group("FIRST_NUMBER")), float(token.group("SECOND_NUMBER"))))
+    else:
+        return str(mean([float(token.group("FIRST_NUMBER")), float(token.group("SECOND_NUMBER"))]))
+
+def _handle_power(token) -> str:
+    return str(float(token.group("FIRST_NUMBER")) ** float(token.group("SECOND_NUMBER")))
+
+def _handle_multiplication_division(token) -> str:
+    if token.group("MULTIPLICATION"):
+        return str(float(token.group("FIRST_NUMBER")) * float(token.group("SECOND_NUMBER")))
+    else:
+        return str(float(token.group("FIRST_NUMBER")) / float(token.group("SECOND_NUMBER")))
+
+def _handle_addition_subtraction(token) -> str:
+    if token.group("ADDITION"):
+        return str(float(token.group("FIRST_NUMBER")) + float(token.group("SECOND_NUMBER")))
+    else:
+        return str(float(token.group("FIRST_NUMBER")) - float(token.group("SECOND_NUMBER")))
```

Comment:
All those function are the same except the operation (*,/,-,+).
Make it generic and avoid duplicated code.